### PR TITLE
♻️ [devext] revamp communication between devtools and background script

### PR DESCRIPTION
**Pull request by BenoitZugmeyer**
## Motivation

Currently, we have two different ways to communicate between the
devtools and background scripts:

* "actions", which broadcast messages to all devtools instances, which
  was created initially to synchronize the store

* "backgroundConnection", which is a persistent connection between
  devtools instances and background and was created to push many events
  from the SDK to the devtools.

"actions" is a bit more limited. Let's make "backgroundConnection"
usable in other contexts, and remove "actions".

## Changes

* Introduce an `EventListeners` class to easily subscribe to various
  communication events

* Make "backgroundConnection" extendable and usable by other modules

* Remove "actions"

* [bonus] reconnect to background automatically

## Testing

- [ ] Local
- [ ] Staging
- [ ] Unit
- [ ] End to end

---

I have gone over the [contributing](https://github.com/DataDog/browser-sdk/blob/main/CONTRIBUTING.md) documentation.

### CODE REVIEW

**Comment by BenoitZugmeyer** [comment]
Current dependencies on/for this PR:
* main
  * **PR #2247** <a href="https://app.graphite.dev/github/pr/DataDog/browser-sdk/2247" target="_blank">Graphite</a> 
    * **PR #2249** <a href="https://app.graphite.dev/github/pr/DataDog/browser-sdk/2249" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/DataDog/browser-sdk/2249?utm_source=stack-comment).

File: developer-extension/src/background/devtoolsPanelConnection.ts [hunk 1]
```diff
@@ -22,6 +23,12 @@ export function sendMessageToDevtools(tabId: number, message: any) {
   port.postMessage(message)
 }
 
+export function sendMessageToAllDevtools(message: BackgroundToDevtoolsMessage) {
```
Comment (BenoitZugmeyer) [inline developer-extension/src/background/devtoolsPanelConnection.ts:26]: Note: this is something that I want to remove in a future PR

**Review by amortemousque — APPROVED** [review]

File: developer-extension/src/background/devtoolsPanelConnection.ts [hunk 2]
```diff
@@ -0,0 +1,59 @@
+import type { BackgroundToDevtoolsMessage, DevtoolsToBackgroundMessage } from '../common/types'
+import { EventListeners } from '../common/eventListeners'
+import { createLogger } from '../common/logger'
+
+const logger = createLogger('devtoolsPanelConnection')
+
+const devtoolsConnections = new Map<number, chrome.runtime.Port>()
```
Comment (bcaudan) [inline developer-extension/src/background/devtoolsPanelConnection.ts:7]: **🥜 nitpick:** naming suggestion‏
```suggestion
const devtoolsConnectionByTabId = new Map<number, chrome.runtime.Port>()
```

File: developer-extension/src/background/devtoolsPanelConnection.ts [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+const portNameRe = /^devtools-panel-for-tab-(\d+)$/
+
+export const onDevtoolsFirstConnection = new EventListeners<number>()
+export const onDevtoolsLastDisconnection = new EventListeners<number>()
```
Comment (bcaudan) [inline developer-extension/src/background/devtoolsPanelConnection.ts:12]: **💭 thought:** ‏a `type TabId = number` alias could make those types more expressives

File: developer-extension/src/panel/sessionReplayPlayer/startSessionReplayPlayer.ts [hunk 4]
```diff
@@ -50,20 +50,25 @@ export function startSessionReplayPlayer(
...
-      } else {
+  const backgroundMessageSubscription = onBackgroundMessage.subscribe((backgroundMessage) => {
+    if (backgroundMessage.type !== 'sdk-message') {
+      return
+    }
+    const sdkMessage = backgroundMessage.message
+    if (sdkMessage.type !== 'record') {
+      return
+    }
+    const record = sdkMessage.payload.record
```
Comment (bcaudan) [inline developer-extension/src/panel/sessionReplayPlayer/startSessionReplayPlayer.ts:61]: **🥜 nitpick:** ‏simpler alternative
```suggestion
    if (!(backgroundMessage.type === 'sdk-message' && backgroundMessage.message.type === 'record')) {
      return
    }
    const record = backgroundMessage.message.payload.record
```

**Review by bcaudan — APPROVED** [review]

**Pull request merged**